data/src/main/java/org/apache/iceberg/MetricsAppender.java

```diff
@@ -0,0 +1,839 @@
+  }
+
+  private enum RecordRepresentation {
+    RECORD,
+    GENERIC_RECORD
+  }
+
+  private enum FixedTypeRepresentation {
+    BYTE_ARRAY,
+    BYTE_BUFFER,
+    GENERIC_FIXED
+  }
+
+  private enum DateTypeRepresentation {
+    INT,
+    LOCAL_DATE
+  }
+
+  private enum TimeTypeRepresentation {
+    LONG,
+    LOCAL_TIME
+  }
+
+  private enum TimestampTypeRepresentation {
+    LONG,
+    OFFSET_DATETIME,
+    LOCAL_DATETIME
+  }
+
+  private static class MetricsAppenderConfiguration {
+    private final FixedTypeRepresentation fixedTypeRepresentation;
+    private final DateTypeRepresentation dateTypeRepresentation;
+    private final TimeTypeRepresentation timeTypeRepresentation;
+    private final TimestampTypeRepresentation timestampTypeRepresentation;
+    private final RecordRepresentation recordRepresentation;
+
+    MetricsAppenderConfiguration(RecordRepresentation recordRepresentation,
+                                 FixedTypeRepresentation fixedTypeRepresentation,
+                                 DateTypeRepresentation dateTypeRepresentation,
+                                 TimeTypeRepresentation timeTypeRepresentation,
+                                 TimestampTypeRepresentation timestampTypeRepresentation) {
+      this.recordRepresentation = recordRepresentation;
+      this.fixedTypeRepresentation = fixedTypeRepresentation;
+      this.dateTypeRepresentation = dateTypeRepresentation;
+      this.timeTypeRepresentation = timeTypeRepresentation;
+      this.timestampTypeRepresentation = timestampTypeRepresentation;
+    }
+
+    public FixedTypeRepresentation getFixedTypeRepresentation() {
+      return fixedTypeRepresentation;
+    }
+
+    public DateTypeRepresentation getDateTypeRepresentation() {
+      return dateTypeRepresentation;
+    }
+
+    public TimestampTypeRepresentation getTimestampTypeRepresentation() {
+      return timestampTypeRepresentation;
+    }
+  }
+
+  public static class Builder<D> {
+
+    private final FileAppender<D> appender;
+    private final Schema schema;
+    private FixedTypeRepresentation fixedTypeRepresentation;
+    private DateTypeRepresentation dateTypeRepresentation;
+    private TimeTypeRepresentation timeTypeRepresentation;
+    private TimestampTypeRepresentation timestampTypeRepresentation;
+    private RecordRepresentation recordRepresentation;
+
+    public Builder(FileAppender<D> appender, Schema schema) {
+      this.appender = appender;
+      this.schema = schema;
+      // TODO default null values or force setting every time?
+      fixedTypeRepresentation = FixedTypeRepresentation.BYTE_ARRAY;
+      dateTypeRepresentation = DateTypeRepresentation.LOCAL_DATE;
+      timeTypeRepresentation = TimeTypeRepresentation.LOCAL_TIME;
+      timestampTypeRepresentation = TimestampTypeRepresentation.OFFSET_DATETIME;
+      recordRepresentation = RecordRepresentation.RECORD;
+    }
+
+    // TODO or setRecordRepresentation(RecordRepresentation) for every option?
```

Comment:
It is probably easier to configure the object model being used, like `avroGenerics`, `icebergGenerics`, `pigTuples`, or `hiveRows`. That's the approach we took to configure the MR InputFormat [object model](https://github.com/apache/iceberg/blob/master/mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java#L189-L197).

That would give you a set of representations:
```java
  public Builder avroGenerics() {
    fixedTypeRepresentation = GENERIC_FIXED;
    recordRepresentation = GENERIC_RECORD;
    dateRepresentation = INT;
    ...
  }
```

But it may still be simpler to add metrics collection to the [`ValueWriter` interface](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/ValueWriter.java), since you could return no metrics by default and add to each implementation, which already has the typed value to write.

Reply:
@rdblue second time you mention the possibility of implement this functionality through ValueWriter. Looks like you prefer that option. Shall I pivot to that solution?

Reply:
That's what I'd recommend. You could do it this way, but I think it's going to be more complexity and work. Sorry, I don't mean to be prescriptive here, it's just that I think you'd have the easiest time that way.